homeassistant/components/roon/media_player.py

```diff
@@ -0,0 +1,521 @@
+"""
+MediaPlayer platform for Roon component.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.roon/
+"""
+import logging
+
+from homeassistant.components.media_player import MediaPlayerEntity
+from homeassistant.components.media_player.const import (
+    SUPPORT_NEXT_TRACK,
+    SUPPORT_PAUSE,
+    SUPPORT_PLAY,
+    SUPPORT_PLAY_MEDIA,
+    SUPPORT_PREVIOUS_TRACK,
+    SUPPORT_SEEK,
+    SUPPORT_SELECT_SOURCE,
+    SUPPORT_SHUFFLE_SET,
+    SUPPORT_STOP,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+)
+from homeassistant.const import (
+    DEVICE_DEFAULT_NAME,
+    STATE_IDLE,
+    STATE_OFF,
+    STATE_PAUSED,
+    STATE_PLAYING,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.util import convert
+from homeassistant.util.dt import utcnow
+
+from .const import DOMAIN
+
+DEPENDENCIES = ["roon"]
+
+SUPPORT_ROON = (
+    SUPPORT_PAUSE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_STOP
+    | SUPPORT_PREVIOUS_TRACK
+    | SUPPORT_NEXT_TRACK
+    | SUPPORT_SHUFFLE_SET
+    | SUPPORT_SEEK
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_VOLUME_MUTE
+    | SUPPORT_PLAY
+    | SUPPORT_PLAY_MEDIA
+    | SUPPORT_SELECT_SOURCE
+    | SUPPORT_VOLUME_STEP
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Roon MediaPlayer from Config Entry."""
+    roon_server = hass.data[DOMAIN][config_entry.data["host"]]
+    media_players = {}
+
+    @callback
+    def async_update_media_player(player_data):
+        """Add or update Roon MediaPlayer."""
+        dev_id = player_data["dev_id"]
+        if dev_id not in media_players:
+            # new player!
+            media_player = RoonDevice(roon_server, player_data)
+            media_players[dev_id] = media_player
+            async_add_entities([media_player])
+        else:
+            # update existing player
+            media_player = media_players[dev_id]
+            if media_player and media_player.entity_id:
+                media_player.update_data(player_data)
+                media_player.async_update_callback(media_player.unique_id)
```

Comment:
Don't access the entity from outside the entity. Signal the entity with the dispatch helper that we already have in place instead.